proposals/2024-06-01-Self-Identification-and-Bundling.md

```diff
@@ -0,0 +1,123 @@
+# Self-Identification and Bundling
+
+## Metadata
+
+|Tag |Value |
+|---- | ---------------- |
+|Proposal |[2024-06-01 Self-Identification-and-Bundling](https://github.com/OAI/OpenAPI-Specification/tree/main/proposals/{2024-06-01-Self-Identification-and-Bundling.md})|
+|Relevant Specification(s)|OAS, Arazzo|
+|Authors|[Henry Andrews](https://github.com/handrews)|
+|Review Manager | TBD |
+|Status |Proposal|
+|Implementations |n/a|
+|Issues |[{issueid}](https://github.com/OAI/OpenAPI-Specification/issues/{IssueId})|
+|Previous Revisions |[{revid}](https://github.com/OAI/OpenAPI-Specification/pull/{revid}) |
+
+## Change Log
+
+|Date |Responsible Party |Description |
+|---- | ---------------- | ---------- |
+|2024-06-01 | @handrews | Initial submission
+
+## Introduction
+
+Poor support for external references has fractured the OAS tooling landscape, with many tools requiring multi-document OpenAPI Descriptions (OADs) to be combined into a single document.  Arazzo requires resolving sources and runtime expressions from multiple OADs, each of which might consist of multiple documents.  There is no way to combine all of the OAD and Arazzo documents involved into a single document, but an alternate solution would be a bundle similar to [what we recommend on our blog for bundling Schema Objects](https://www.openapis.org/blog/2021/08/23/json-schema-bundling-finally-formalised).  This would require a similar mechanism to JSON Schema's `$id` for OAS and Arazzo documents and the components within them.  It would also provide an alternative to current multi-to-single-document OAD tools, most (possibly all) of which do not fully support OAS 3.1, and allow for _lossless bundling of identifiable components_, which is increasingly needed by industry standards groups publishing API "building blocks" for use across many APIs by many different providers.
```

Comment:
> There is no way to combine all of the OAD and Arazzo documents involved into a single document

I'd like to understand why this is the case. What about just OAD documents?

Reply:
Objects in descriptions are scoped to the document in which they originate. We don't really have semantics for merging multiple documents. In some cases, it would involve renaming unique identifiers like map keys. Tools that have done this create an inconsistent experience, and to reverse the process requires a source map of some kind. Maintaining document identity in a multi-document scenario obviates the need for that.

Reply:
@mikekistler basically what @kevinswiber said, but I'll add two things:

I chatted with someone on some Slack a while ago who was having a lot of problems because his "merge to a single document" tools that he needed in order to get something (AWS gateway, maybe?) to accept the OAD was messing things up with what it though were "safe" transformations.  As Kevin pointed out, there are a lot of choices that tools doing this have to make, and those can be surprising and sometimes breaking.  When you start with a single document, split it, and then use the same toolchain to re-combine the pieces, that works quite well because the split and combine tool are the same and make the same assumptions.  But mixing toolchains doesn't work well here.

Another way to look at it is to look at how [JSON Schema bundling works](https://www.openapis.org/blog/2021/08/23/json-schema-bundling-finally-formalised).  It relies heavily on `"$id"` to preserve not just the referencing behavior but the literal reference values.  This way there isn't any "rewriting" of the documents, and no need to merge them.  They are simply placed in the standard location (`"$defs"`) and used in a way that the name under `"$defs"` is irrelevant.

As briefly mentioned in the call this morning, if we want to support `"$id"` (or an equivalent named differently to reduce confusion with JSON Schema) in every Object type in the OAS and Arazzo, we could absolutely implement single-document bundling.  But nested `"$id"`s make it much harder to determine the base URI, which is something that tooling vendors seem to struggle with (or are just indifferent to) as it is.  Using a document stream minimizes the changes and implementation work involved.

I'm trying to get the maximum improvement for the minimum effort with this proposal.